File: 2022/P04/problem4a.py
data = []
with open("problem4.txt", 'r') as f:
    data = f.read().strip('\n').split('\n')

total = 0
for d in data:
    p = d.split(',')
    l1, h1 = [int(x) for x in p[0].split('-')]
    l2, h2 = [int(x) for x in p[1].split('-')]
    if l1 >= l2 and h1 <= h2:
        total += 1
    elif l2 >= l1 and h2 <= h1:
        total += 1

print(total)

File: 2023/P04/problem4a.rs
use std::fs;

fn main() {
    let data = fs::read_to_string("problem4.txt")
        .expect("Failed to read file");
    let data = data.trim()
        .split('\n')
        .collect::<Vec<_>>();
    let mut sum = 0;
    for line in data {
        let card = line.split(": ").collect::<Vec<_>>()[1];
        let vec = card.split(" | ").collect::<Vec<_>>();
        let wins = vec[0].split(" ").collect::<Vec<_>>();
        let wins = wins.iter().filter(|&&x| x != "").collect::<Vec<_>>();
        let mine = vec[1];
        let mut val = 0;
        for v in mine.split(" ") {
            match wins.iter().find(|&&x| x == &v) {
                Some(_) => {
                    if val == 0 {
                        val = 1;
                    } else {
                        val *= 2;
                    }
                },
                None => {}
           }
        }
        sum += val;
    }
    println!("{sum}");
}
